### DISCUSSION TITLE
Como criar de forma estruturada um Domínio?

### DISCUSSION THREAD

**Discussion by RafaelAlmeid00** [RafaelAlmeid00/API#20] [Ideas]
# Como criar um Domínio de forma estruturada

Este guia fornece instruções passo a passo sobre como criar um Domínio de forma estruturada em um projeto .NET Core.

## 1. Defina as notações na Classe de Domínio

No diretório `src/Domain/Models`, você encontrará as classes de domínio. É importante definir as propriedades como possivelmente nulas para permitir a pesquisa por qualquer propriedade no controller pela função `Search(Domain data)`.

Exemplo:

```csharp
public record ListCpf : IListCpfDTO
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int? ListId { get; set; }
    [ForeignKey("Business")]
    public string? BusinessCnpj { get; set; }
    public string? ListType { get; set; }
    public string? ListCPF { get; set; }
    public virtual Business? BusinessCnpjNavigation { get; set; }
    public virtual ICollection<User>? Users { get; set; } = [];
}
```

## 2. Implemente a interface de Domínio

No diretório src/Interfaces/Models, você encontrará as interfaces de domínio. Implemente essas interfaces nas suas classes de domínio.

Exemplo:

```csharp
public interface IListCpfDTO
{
    int? ListId { get; set; }
    string? BusinessCnpj { get; set; }
    string? ListType { get; set; }
    string? ListCPF { get; set; }
    Business? BusinessCnpjNavigation { get; set; }
    ICollection<User>? Users { get; set; }
}
```

## 3. Crie a interface de Service e Repository do Domínio

No diretório src/Interfaces, crie as interfaces de Service e Repository para o seu domínio.

Exemplo:

```csharp
public interface IServiceListCpf : IBaseService<ListCpf>, 
    IDeleteEditService<ListCpf>, 
    IAlterTypeService<ListCpf>
{
    Task<IResultadoOperacao<dynamic>> CreateWithExcel(IFormFile data);
}

public interface IRepositoryListCpf : IBaseRepository<ListCpf>, 
    IDeleteEditRepository<ListCpf>, 
    IAlterTypeRepository<ListCpf>
{
    Task<IResultadoOperacao<dynamic>> CreateWithExcel((List<dynamic>, List<dynamic>) data);
}
```

## 4. Crie o controller injetando as interfaces genéricas

No diretório src/Adapters_Controllers, crie o controller para o seu domínio e injete as interfaces genéricas.

Exemplo:

```csharp
namespace Api.Adapters_Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class ListCpfController : ControllerBase
    {
        private readonly IServiceListCpf _service;
        private readonly IAntiforgery _antiforgery;

        public ListCpfController(IServiceListCpf service, IAntiforgery antiforgery)
        {
            _service = service;
            _antiforgery = antiforgery;
        }

        [Authorize]
        [RequireAntiforgeryToken]
        [HttpGet]
        public async Task<ActionResult> GetListCpfs([FromBody] ListCpf data)
        {
            // Implemente a lógica para obter a lista de CPFs
        }
    }
}
```

## 5. Crie a classe de serviço

No diretório src/Services, crie a classe de serviço para o seu domínio e implemente as manipulações de dados pertinentes.

Exemplo:

```csharp

public class ServiceListCpf : IServiceListCpf
{
    private readonly IRepositoryListCpf _repository;
    private readonly IListCPFValidator _validator;

    public ServiceListCpf(IRepositoryListCpf repository, IListCPFValidator validator)
    {
        _repository = repository;
        _validator = validator;
    }

    // Implemente as funções específicas do serviço aqui
}
```

## 6. Crie o Repository

No diretório src/Repositories, crie o Repository para o seu domínio e implemente a interface Repository criada anteriormente.

Exemplo:

```csharp
public class ListCpfRepository : IRepositoryListCpf
{
    private readonly EasyPassContext _context;

    public ListCpfRepository(EasyPassContext context)
    {
        _context = context;
    }

    // Implemente as operações de CRUD aqui
}
```

## 7. Siga o padrão de resposta HATEOAS

Implemente o padrão de resposta HATEOAS para suas operações.

Exemplo:

```csharp
namespace Api.Domain
{
    public record ResultadoOperacao<T> : IResultadoOperacao<T>
    {
        public bool Sucesso { get; set; }
        public string? Erro { get; set; }
        public T? Data { get; set; }
        public required ILink Link { get; set; }
    }

    public record Link : ILink
    {        
        public required string Rel { get; set; }
        public required string Href { get; set; }
        public required string Method { get; set; }
    }
}

```

## 7. Implemente no Middlware

Implemente as rotas do Domínio no Middlware no diretório `src/Infrastructure`. Importe lembrar que usuário pode fazer GET em alguns Domínios mas não pode alterar nem gerar nenhum dado, somente a Bussines e Admin tem acesso para gerar ou alterar dados que não são do usuário, sendo Bussines restrito aos seus Domínios da suas regras de negócio apenas.

Exemplo:

```csharp
 if (context.Request.Path.StartsWithSegments("/api/ListCpf") 
     && BussinesClaim is null && AdminClaim is null 
     && context.Request.Method != "GET")
{
   context.Response.StatusCode = StatusCodes.Status401Unauthorized;
   await WriteUnauthorizedResponse(context);
   return;
}
```

**Comment by LuisEduardoSpindola** [comment]
Nota: Não esquecer de verificar a injeção de dependência também
